File: pathtosum2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
// Time Complexity : O(N) - > visit every node
// Space Complexity : O(N) - > stack
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
"""
class Solution:
    def pathSum(self, root, sum):
        
        self.result = []
        self.dfs(root, sum, [])
        return self.result
    
    def dfs(self, root, rsum, stack):
        
        if not root:
            return
        
        stack.append(root.val)
        if root.left is None and root.right is None:                
            if rsum == root.val:
                self.result.append(list(stack))
                    
        self.dfs(root.left, rsum - root.val, stack)        
        self.dfs(root.right, rsum - root.val, stack)
        
        stack.pop()
        
        

File: pathtosum2.java
//TC:O(n^2)
//SC:O(height*diameter)
//Successfully ran on leetcode
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum)
    {
        List<List<Integer>> output=new ArrayList<>();
        List<Integer> path=new ArrayList<>();

        helper(root,targetSum,0,output,path);

        return output;
    }
    private void helper(TreeNode root,int target,int currSum,List<List<Integer>> output,List<Integer> path)
    {
        if(root==null)
        {
            return;
        }
        currSum+=root.val;
        path.add(root.val);

        if(root.left==null && root.right==null)
        {
            if(currSum==target)
            {

                output.add(new ArrayList<>(path));
            }
            path.remove(path.size()-1);
            return;
        }
        helper(root.left,target,currSum,output,path);
        helper(root.right,target,currSum,output,path);

        path.remove(path.size()-1);
    }
}
